### PR TITLE
Mongodb geoserver plugin

### PR DESCRIPTION
**Pull request by andrewjstrominger**

### CODE REVIEW

**Comment by andrewjstrominger** [comment]
dbwfsschema-no-details
dbwfsschema

File: src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+            subCollectionAttributes.each { subCollectionAttribute ->
+                subCollectionAttribute.subCollectionPath = "${subCollection.subCollectionPath}${subCollectionAttribute.subCollectionPath ? '.' + subCollectionAttribute.subCollectionPath : ''}".toString()
+                attributes << subCollectionAttribute
+            }
+        }
+        attributes.unique()
+        return attributes
+    }
+
+    /*BasicDBObject getPropertyIsEqualToQuery(String propertName, String id) {
```
Comment (andrewjstrominger) [inline src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy]: remove this

File: src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    @Override
+    SimpleFeatureCollection subCollection(Filter filter) {
+        return  // TODO
+    }
+
+    @Override
+    SimpleFeatureCollection sort(SortBy order) {
+        log.info("order = ${order}")
```
Comment (andrewjstrominger) [inline src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy]: remove this

File: src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+
+        DBCursor dbCursor = this.dbCollection.find(dbQuery)
+        if(supportsMaxAndOffsetQueries() &&  query?.getMaxFeatures() != null) {
+            dbCursor?.limit(query.getMaxFeatures())
+        }
+
+        if (supportsMaxAndOffsetQueries() && query?.getStartIndex() != null) {
+            dbCursor?.skip(query.getStartIndex())
+        }
+        //log.info "dbQuery = ${dbQuery}"
```
Comment (andrewjstrominger) [inline src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy:186]: remove this

File: src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+        }
+
+        value = convertValueIfNeeded(propertyName, value)
+        def propQuery = mongoOperator == null ? value : new BasicDBObject(mongoOperator, value)
+
+        return new BasicDBObject(propertyName, propQuery)
+    }
+
+    protected def convertValueIfNeeded(String dbQueryPath, String valueString) {
+        def convertedValue = valueString // TODO
```
Comment (andrewjstrominger) [inline src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy]: TODO

File: src/test/groovy/com/spidasoftware/mongodb/feature/iterator/MongoDBFeatureIteratorSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.spidasoftware.mongodb.feature.iterator
+
+import spock.lang.Specification
+
+class MongoDBFeatureIteratorSpec extends Specification {
```
Comment (andrewjstrominger) [inline src/test/groovy/com/spidasoftware/mongodb/feature/iterator/MongoDBFeatureIteratorSpec.groovy]: Implement

File: src/main/groovy/com/spidasoftware/mongodb/data/MongoDBFeatureSource.groovy [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            if(coordinate != null) {
+                referencedEnvelope.expandToInclude(coordinate)
+            }
+        }
+        return referencedEnvelope
+    }
+
+    @Override
+    public int getCount(Query query) throws IOException {
+        return dbCollection.find().size() // TODO
```
Comment (spidadeveloper) [inline src/main/groovy/com/spidasoftware/mongodb/data/MongoDBFeatureSource.groovy]: TODO

File: src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+            // If targetCRS is DefaultProjectedCRS then transform doesn't change lng, lat to lat, lng so check if it should be changed
+            if (this.targetCRS instanceof DefaultProjectedCRS && this.targetCRS?.baseCRS?.coordinateSystem?.getAxis(0)?.direction == AxisDirection.NORTH) {
+                point = new GeometryFactory().createPoint(new Coordinate(point.getCoordinate().y, point.getCoordinate().x))
+            }
+
+            point.setUserData(this.targetCRS)
+
+            GeometryDescriptor geometryDescriptor = featureType.getGeometryDescriptor()
+            if (this.targetCRS != this.sourceCRS) {
+                ComplexFeatureTypeFactoryImpl complexFeatureTypeFactory = new ComplexFeatureTypeFactoryImpl()
```
Comment (spidadeveloper) [inline src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy:193]: Maybe remove?

File: src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    @Override
+    String getID() {
+        return null
+    }
+
+    @Override
+    void accepts(FeatureVisitor visitor, ProgressListener progress) throws IOException {
+        // TODO
```
Comment (spidadeveloper) [inline src/main/groovy/com/spidasoftware/mongodb/feature/collection/AbstractMongoDBFeatureCollection.groovy]: remove todo

File: src/main/groovy/com/spidasoftware/mongodb/feature/collection/MongoDBSubCollectionFeatureCollection.groovy [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                return getKeyAndValueFeatures(key, value)
+            }.flatten()
+        }
+    }
+
+    private List<Feature> getUseObjectKeyFeatures(Map attributes, DBObject dbObject, BasicDBObject objectMapping, Object subCollectionObject = null, Integer index = null) {
+        if(!(subCollectionObject ?: dbObject).any { key, value -> value instanceof BasicDBObject}) {
+            return []
+        }
+        def userObjectKeyAttr = objectMapping.attributes.find { it.useObjectKey }
```
Comment (spidadeveloper) [inline src/main/groovy/com/spidasoftware/mongodb/feature/collection/MongoDBSubCollectionFeatureCollection.groovy:129]: figure out what this is and document it

File: src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+        return mapping.subCollections == null || mapping.subCollections.size() == 0
+    }
+
+    FeatureCollection getFeatureCollection(Query query = null) {
+        def andQuery = new BasicDBList()
+        def defaultCollectionQueries = getDefaultCollectionQueries(mapping)
+        def dbQuery = defaultCollectionQueries
+        def filter = query?.getFilter() ?: Filter.INCLUDE
+        def queryFromFilter
+        if(filter != null) {
```
Comment (spidadeveloper) [inline src/main/groovy/com/spidasoftware/mongodb/filter/FilterToDBQuery.groovy]: would never be null

File: src/test/groovy/com/spidasoftware/mongodb/feature/collection/MongoDBSubCollectionFeatureCollectionIteratorSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,836 @@
...
+import org.opengis.feature.Feature
+import org.opengis.feature.type.FeatureType
+
+import org.opengis.filter.Filter
+import spock.lang.Shared
+import spock.lang.Specification
+import org.geotools.filter.text.cql2.CQL
+import java.util.logging.Logger
+
+class MongoDBSubCollectionFeatureCollectionIteratorSpec extends Specification {
```
Comment (spidadeveloper) [inline src/test/groovy/com/spidasoftware/mongodb/feature/collection/MongoDBSubCollectionFeatureCollectionIteratorSpec.groovy]: name?

File: src/test/groovy/com/spidasoftware/mongodb/feature/iterator/MongoDBFeatureIteratorSpec.groovy (hunk 5 above)
Comment (spidadeveloper) [reply src/test/groovy/com/spidasoftware/mongodb/feature/iterator/MongoDBFeatureIteratorSpec.groovy]: Implement me

**Pull request merged**